### PR TITLE
PowerVS: Add DNS Record and VPC LB deletion

### PR DESCRIPTION
**Pull request by dharaneeshvrd**
**What this PR does / why we need it**:

Current powervs destroy infra does not have DNS record deletion and Load balancer deletion in VPC subnet.
This PR implements the same to delete them.

Introduced `BaseDomain, CisCRN and CisDomainID` as part of `PowerVSPlatformDestroyOptions` so when we call destroy infra from destroy cluster, CIS domain details can be passed to attempt the DNS record deletion.

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: cmd/infra/powervs/destroy.go [hunk 1]
```diff
@@ -533,9 +597,18 @@ func destroyVpcSubnet(options *DestroyInfraOptions, infra *Infra, resourceGroupI
 	return pagingHelper(f)
 }
 
-// deleteVpcSubnet deletes the subnet id passed
-func deleteVpcSubnet(id string, v1 *vpcv1.VpcV1, infraID string) (err error) {
-	log(infraID).Info("Deleting VPC subnet", "subnetId", id)
+// deleteVpcSubnet deletes the subnet id passed and LB attached to it
+func deleteVpcSubnet(id string, v1 *vpcv1.VpcV1, options *DestroyInfraOptions) (err error) {
+	// deleting the load balancer before proceeding to subnet deletion, since LB is an attached resource to the subnet
+	e := destroyVpcLB(options, id, v1)
```
Comment (enxebre) [inline cmd/infra/powervs/destroy.go:603]: e?

File: cmd/infra/powervs/destroy.go [hunk 2]
```diff
@@ -533,9 +597,18 @@ func destroyVpcSubnet(options *DestroyInfraOptions, infra *Infra, resourceGroupI
...
+	if e != nil {
+		log(options.InfraID).Error(e, "error destroying VPC Load Balancer")
+		err = fmt.Errorf("error destroying VPC Load Balancer %w", e)
```
Comment (enxebre) [inline cmd/infra/powervs/destroy.go:606]: to be consistent with the rest of the code base, can we just return directly `return fmt.Errorf("error destroying VPC Load Balancer %w", e)`

File: cmd/infra/powervs/destroy.go [hunk 3]
```diff
@@ -555,3 +628,77 @@ func deleteVpcSubnet(id string, v1 *vpcv1.VpcV1, infraID string) (err error) {
...
+	}
+
+	return pagingHelper(f)
+}
+
+// deleteVpcLB deletes the load balancer id passed
+func deleteVpcLB(id string, v1 *vpcv1.VpcV1, infraID string) (err error) {
+	log(infraID).Info("Deleting VPC LoadBalancer:", "id", id)
+	_, err = v1.DeleteLoadBalancer(&vpcv1.DeleteLoadBalancerOptions{ID: &id})
+
```
Comment (enxebre) [inline cmd/infra/powervs/destroy.go:687]: this if could be inlined

File: cmd/cluster/core/destroy.go [hunk 4]
```diff
@@ -47,7 +47,10 @@ type AzurePlatformDestroyOptions struct {
 }
 
 type PowerVSPlatformDestroyOptions struct {
+	BaseDomain    string
 	ResourceGroup string
+	CisCRN        string
```
[the remark this answers is not in the document]
Comment (dharaneeshvrd) [reply cmd/cluster/core/destroy.go]: renamed.

File: cmd/infra/powervs/destroy.go (hunk 1 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go]: modified

File: cmd/infra/powervs/destroy.go (hunk 2 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go]: returned the error directly.

File: cmd/infra/powervs/destroy.go (hunk 3 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go]: modified.

File: cmd/infra/powervs/create.go [hunk 5]
```diff
@@ -326,7 +325,8 @@ func (infra *Infra) setupBaseDomain(options *CreateInfraOptions) (err error) {
 	}
 
 	if rcv2 == nil {
-		return fmt.Errorf("unable to get resource controller")
+		err = fmt.Errorf("unable to get resource controller")
```
Comment (enxebre) [inline cmd/infra/powervs/create.go:328]: why this change? the previous is consistent with the style across the code base.

File: cmd/infra/powervs/create.go [hunk 6]
```diff
@@ -314,9 +314,8 @@ func getIAMAuth() *core.IamAuthenticator {
 	}
 }
 
-// setupBaseDomain get domain id and crn of given base domain
-// TODO(dharaneeshvrd): Currently, resource group provided will be considered only for VPC and PowerVS. Need to look at utilising a common resource group in future for CIS service too and use it while filtering the list
-func (infra *Infra) setupBaseDomain(options *CreateInfraOptions) (err error) {
+// getCisDomainDetails getting CIS domain details like CRN and domainID
+func getCisDomainDetails(baseDomain string) (CisCRN string, CisDomainID string, err error) {
```
Comment (enxebre) [inline cmd/infra/powervs/create.go:318]: is there a reason to use by named return variables? it's not consistent with the rest of the code base of the project and IMO makes readability less obvious than explicit returns.

**Comment by enxebre** [comment]
thanks! can we please align with the rest of the code base be explicit on returns and only use named return variables if there's a particular reason for it.

File: cmd/infra/powervs/destroy.go [hunk 7]
```diff
@@ -555,3 +626,75 @@ func deleteVpcSubnet(id string, v1 *vpcv1.VpcV1, infraID string) (err error) {
...
+		isDone = true
+		return
+	}
+
+	return pagingHelper(f)
+}
+
+// deleteVpcLB deletes the load balancer id passed
+func deleteVpcLB(id string, v1 *vpcv1.VpcV1, infraID string) (err error) {
```
Comment (mkumatag) [inline cmd/infra/powervs/destroy.go:682]: don't we need to delete the backend pool before deleting the load balancer? will it allow you to delete if backend pool exist?

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: cmd/infra/powervs/create.go (hunk 6 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go]: Yes I can see on rest of the code base not using named return variables. Will refactor the code to follow the same.

File: cmd/infra/powervs/create.go [hunk 8]
```diff
@@ -396,7 +396,18 @@ func (infra *Infra) setupBaseDomain(options *CreateInfraOptions) (err error) {
...
+}
+
+// setupBaseDomain get domain id and crn of given base domain
+// TODO(dharaneeshvrd): Currently, resource group provided will be considered only for VPC and PowerVS. Need to look at utilising a common resource group in future for CIS service too and use it while filtering the list
+func (infra *Infra) setupBaseDomain(options *CreateInfraOptions) (err error) {
+
+	infra.CisCrn, infra.CisDomainID, err = getCisDomainDetails(options.BaseDomain)
+
 	if err != nil {
+		err = fmt.Errorf("error retrieving cis domain details %w", err)
```
[the remark this answers is not in the document]
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go]: modified.

File: cmd/infra/powervs/destroy.go [hunk 9]
```diff
@@ -206,6 +222,54 @@ func (options *DestroyInfraOptions) DestroyInfra(infra *Infra) (err error) {
...
+
+	if options.CisCRN == "" || options.CisDomainID == "" {
+		if options.BaseDomain == "" || options.Name == "" {
+			log(options.InfraID).Info("Skipping DNS record deletion. DNS Record deletion requires --name & --base-domain")
+			return nil
+		} else {
+			var err error
+			options.CisCRN, options.CisDomainID, err = getCisDomainDetails(options.BaseDomain)
+			if err != nil {
+				err = fmt.Errorf("error retrieving cis domain details %w", err)
```
[the remark this answers is not in the document]
Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go]: modified

**Comment by dharaneeshvrd** [comment]
> thanks! can we please align with the rest of the code base be explicit on returns and only use named return variables if there's a particular reason for it.

I have modified the code for the above comment with respect to the changes made in this commit. For the whole code base, can we keep it in different PR?

File: cmd/infra/powervs/destroy.go [hunk 10]
```diff
@@ -206,6 +222,54 @@ func (options *DestroyInfraOptions) DestroyInfra(infra *Infra) (err error) {
...
+		return err
+	}
+
+	if len(dnsRecordsL.Result) <= 0 {
+		log(options.InfraID).Info("No matching DNS Records present in CIS Domain")
+		return nil
+	}
+
+	record := dnsRecordsL.Result[0]
+	if *record.Name == recordName {
```
[the remark this answers is not in the document]
Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go]: Sry this is not required, modified.

File: cmd/infra/powervs/destroy.go (hunk 7 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go:678]: No Manju, deleting the load balancer will take care of removing backend pool and listeners.

File: cmd/infra/powervs/destroy.go [hunk 11]
```diff
@@ -206,6 +222,50 @@ func (options *DestroyInfraOptions) DestroyInfra(infra *Infra) (err error) {
...
+
+	recordName := fmt.Sprintf("*.apps.%s.%s", options.Name, options.BaseDomain)
+	listDnsRecordsOpt := &dnsrecordsv1.ListAllDnsRecordsOptions{Name: &recordName}
+
+	dnsRecordsL, _, err := dnsRecordsV1.ListAllDnsRecords(listDnsRecordsOpt)
+	if err != nil {
+		return err
+	}
+
+	if len(dnsRecordsL.Result) <= 0 {
```
Comment (mkumatag) [inline cmd/infra/powervs/destroy.go:254]: just wondering if length every be less than 0 🤔

**Review by mkumatag — CHANGES_REQUESTED** [review]
small nit, other lgtm

File: cmd/infra/powervs/destroy.go (hunk 11 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go]: modified to `len(dnsRecordsL.Result) < 1`, does this looks good? @mkumatag

Comment (mkumatag) [reply cmd/infra/powervs/destroy.go]: how about just `len(dnsRecordsL.Result) == 0`?

Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go]: modified🙂 @mkumatag

**Comment by dharaneeshvrd** [comment]
@enxebre Please have a look when you get a chance!

File: cmd/infra/powervs/destroy.go [hunk 12]
```diff
@@ -206,6 +222,50 @@ func (options *DestroyInfraOptions) DestroyInfra(infra *Infra) (err error) {
 	return
 }
 
+// deleteDNSRecords deletes DNS records from CIS domain
+func deleteDNSRecords(options *DestroyInfraOptions) error {
+
+	if options.CisCRN == "" || options.CisDomainID == "" {
+		if options.BaseDomain == "" || options.Name == "" {
```
Comment (mkumatag) [inline cmd/infra/powervs/destroy.go:229]: is there reason why we made these 2 options as optional? if we make them mandatory then we don't need this check!

File: cmd/infra/powervs/destroy.go [hunk 13]
```diff
@@ -206,6 +222,50 @@ func (options *DestroyInfraOptions) DestroyInfra(infra *Infra) (err error) {
...
+		} else {
+			var err error
+			options.CisCRN, options.CisDomainID, err = getCisDomainDetails(options.BaseDomain)
+			if err != nil {
+				return fmt.Errorf("error retrieving cis domain details %w", err)
+			}
+		}
+	}
+
+	dnsRecordsV1, err := dnsrecordsv1.NewDnsRecordsV1(&dnsrecordsv1.DnsRecordsV1Options{Crn: &options.CisCRN, ZoneIdentifier: &options.CisDomainID, Authenticator: getIAMAuth()})
```
Comment (mkumatag) [inline cmd/infra/powervs/destroy.go:241]: There is a possibility of `options.CisCRN, options.CisDomainID` can be empty string, if they are empty then do we really need to proceed further?

case: where user deletes the dns entries manually

File: cmd/infra/powervs/destroy.go [hunk 14]
```diff
@@ -555,3 +622,75 @@ func deleteVpcSubnet(id string, v1 *vpcv1.VpcV1, infraID string) (err error) {
...
+
+		isDone = true
+		return
+	}
+
+	return pagingHelper(f)
+}
+
+// deleteVpcLB deletes the load balancer id passed
+func deleteVpcLB(id string, v1 *vpcv1.VpcV1, infraID string) (err error) {
```
Comment (mkumatag) [inline cmd/infra/powervs/destroy.go:678]: I feel there will be confusion of `deleteVpcLB` vs `destroyVpcLB`, to keep it simple I'm wondering if we can move this code into `destroyVpcLB` as a simple inline function which takes id as arg

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: cmd/infra/powervs/destroy.go (hunk 12 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go]: Yes right Manju, added them as mandatory and removed this check.

File: cmd/infra/powervs/destroy.go (hunk 13 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go]: No we are depending the base-domain's CRN and DomainID only right, if they remove DNS entry, we can still get CRN and DomainID info. 
If they remove base-domain itself, it will be empty, in that case `getCisDomainDetails` will return an error and we won't proceed.

**Review by mkumatag — APPROVED** [review]
/lgtm

File: cmd/infra/powervs/create.go [hunk 15]
```diff
@@ -375,37 +373,48 @@ func (infra *Infra) setupBaseDomain(options *CreateInfraOptions) (err error) {
 
 			if zoneList != nil {
 				for _, zone := range zoneList.Result {
-					if *zone.Name == options.BaseDomain {
-						infra.CisCrn = *resource.CRN
-						infra.CisDomainID = *zone.ID
-						isDone = true
-						return
+					if *zone.Name == baseDomain {
```
Comment (enxebre) [inline cmd/infra/powervs/create.go:376]: is the name a universal id or there might be more than one with same name?

File: cmd/infra/powervs/destroy.go [hunk 16]
```diff
@@ -555,3 +619,68 @@ func deleteVpcSubnet(id string, v1 *vpcv1.VpcV1, infraID string) (err error) {
 	err = wait.PollImmediate(pollingInterval, vpcResourceDeletionTimeout, f)
```
Comment (enxebre) [inline cmd/infra/powervs/destroy.go:619]: nit: this can just return wait.Poll.....

File: cmd/infra/powervs/destroy.go [hunk 17]
```diff
@@ -555,3 +619,68 @@ func deleteVpcSubnet(id string, v1 *vpcv1.VpcV1, infraID string) (err error) {
...
+
+		// For paging over next set of resources getting the start token and passing it for next iteration
+		if loadBalancerL.Next != nil && *loadBalancerL.Next.Href != "" {
+			return false, *loadBalancerL.Next.Href, nil
+		}
+
+		return true, "", nil
+	}
+
+	return pagingHelper(f)
```
Comment (enxebre) [inline cmd/infra/powervs/destroy.go:685]: why does this pagingHelper?

**Comment by enxebre** [comment]
nit: please use imperative mood for commit message rather than pass https://github.com/openshift/hypershift/blob/main/docs/content/contribute/index.md

File: cmd/infra/powervs/create.go (hunk 15 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go:376]: There is no possibility for dup name to exist. Name will be unique under a CIS service ID.

File: cmd/infra/powervs/create.go [hunk 18]
```diff
@@ -314,30 +314,29 @@ func getIAMAuth() *core.IamAuthenticator {
 	}
 }
 
-// setupBaseDomain get domain id and crn of given base domain
-// TODO(dharaneeshvrd): Currently, resource group provided will be considered only for VPC and PowerVS. Need to look at utilising a common resource group in future for CIS service too and use it while filtering the list
-func (infra *Infra) setupBaseDomain(options *CreateInfraOptions) (err error) {
+// getCisDomainDetails getting CIS domain details like CRN and domainID
```
[the remark this answers is not in the document]
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go]: modified.

File: cmd/infra/powervs/destroy.go (hunk 16 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go]: modified.

**Comment by dharaneeshvrd** [comment]
Thanks @enxebre, I think, I have addressed the comments, ptal.
